### PR TITLE
Immutable operations

### PR DESCRIPTION
**Pull request by vineetbansal**
Certain operations on `ImageSource` now return a copy. 

```
cache
downsample
whiten
phase_flip
invert_contrast
normalize_background
```

We're only storing primitive data types in the `metadata` dataframe, which is why the behavior specified at https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.copy.html ("actual Python objects will not be copied recursively..") is not a problem for us. The rest of the references (`unique_filters` etc.) are deepcopied properly as expected.

The behavior is implemented as a simple decorator right now, but can be changed once we're satisfied with the API and have verified its correctness.

Based on a recent discussion with @garrettwrong, I'll implement an `.update` method (method name more or less random and can be changed) on `ImageSource` next, which will allow us to do something like (leaving `avgs` untouched, and using `set_metadata` under the hood):
```
avgs_with_rotations = avgs.update(rotations=rots_est)
```
instead of
```
avgs.rotations = rots_est
```

### CODE REVIEW

**Comment by vineetbansal** [comment]
The reason tests are failing is that `finufft-2.1.0` needs a patch to do a check for value equality instead of reference equality (which was the intention of those checks), something that shows up in `deepcopy`.

Specifically, the four instances of:
```
if x is not None and x.dtype is not np.dtype('
```
in the released version of `_interfaces.py` need to be replaced by:
```
if x is not None and x.dtype != np.dtype('
```

This may not be an issue with their current code, but given that the release was made in June 2022, I'm not sure how to proceed with this. I can open up a PR there and see how it goes.

File: src/aspire/source/image.py
Comment (garrettwrong) [inline src/aspire/source/image.py:102]: I know it was undecided whether we would keep the decorator, but I actually like it... if future developers decide immutable was one of my bad decisions one day, it will be easier to remove.

Anyway, ready to add a docstring. Just something brief about the intent is fine.

File: src/aspire/source/image.py
Comment (garrettwrong) [inline src/aspire/source/image.py:633]: Just curious, was this one needed for anything other than `ArrayImageSource`?

**Review by garrettwrong — CHANGES_REQUESTED** [review]
Code looks clean, thanks! Just one minor string and a question.  

The large merge into `develop` today may have spawned a few more updates.  The last large PR for this release, #882, will probably require some similar updates, but it will likely be after your PR is merged (so in my court :) ).

I do want to review the output of gallery and experiments tomorrow, but I will do that first thing in the morning to approve.

File: gallery/tutorials/preprocess_imgs_sim.py
Comment (garrettwrong) [inline gallery/tutorials/preprocess_imgs_sim.py]: Reviewing the gallery in more detail, do you think this (and the following three preprocessing steps) might be a good place to actually demonstrate the new api is not mutating? Historically I think this particular `downsampling` was an area that raised weird behavior to attention.  Or maybe it should be a second section in the notebook? Like show steps as independent sources, then as sequence of sources, and maybe finally as chained (below).  I would sneak `.cache` after the initial source to make it faster if we go down this road.

It would just require changing the `source =` to `source_nb = ` etc.  The `[0].asnumpy()` can be pushed into the plotting which is already messy. The `[0]` will be redundant with `idm` below anyway.

We can show the ability to chain.

`chained_source = source.phase_flip().normalize_background().whiten(noise_estimator).invert_contrast()`

What do you think?

**Review by garrettwrong — CHANGES_REQUESTED** [review]
Overall gallery looked good. Only one additional remark.

To sync with the merge from yesterday should be just a few more boilerplate changes.

File: gallery/tutorials/preprocess_imgs_sim.py
Comment (garrettwrong) [reply gallery/tutorials/preprocess_imgs_sim.py]: Talked with Vineet, I'll pick this gallery edit up as a follow up task.

**Comment by vineetbansal** [comment]
I tested the `as_copy` behavior against the current HEAD on `finufft` (`a0db2d0b`), and we don't need the hack we're using if we use that version of finufft. So when we update our dependencies, we can remove the entire `__deepcopy__` override for `ImageSource`.

Of course there may be other similar or unrelated hacks we need to do for other nfft backends that I haven't tested, so it may be worthwhile to make a historical note of it somewhere.

File: src/aspire/source/image.py
Comment (vineetbansal) [reply src/aspire/source/image.py:633]: I do see some uses of `cache()` for `RelionSource` objects when executing certain commands. For example:
https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/e3145c1a3c5a9b4a18d3c917e15e20d4569d8088/src/aspire/commands/cov3d.py#L51
https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/e3145c1a3c5a9b4a18d3c917e15e20d4569d8088/src/aspire/commands/denoise.py#L81

I don't think `RelionSource` is related to `ArrayImageSource`. Or perhaps I'm misunderstanding the question..

**Comment by garrettwrong** [comment]
> Of course there may be other similar or unrelated hacks we need to do for other nfft backends that I haven't tested, so it may be worthwhile to make a historical note of it somewhere.

@vineetbansal , cufinufft has the `!=dtype` , so we'll avoid that...  It unfortunately has a different issue in the published package (`is` vs `is not None` iirc). However, the `ampere_gpu` CI job _is_ running the master version of cufinufft with my patch... so your code is running with the other backend too :).

File: src/aspire/source/image.py
Comment (garrettwrong) [reply src/aspire/source/image.py:633]: I don't know that the `cache` needs to make/return a copy. The image array it is storing should be immutable.  Cache is just saying, generate and store stuff inside self now, so it is ready.  In other words, `cache` can probably return itself after making and assigning the cached data array.  I don't personally consider it preprocessing (it isn't mutating anything in the pipeline).

I guess there is an argument that decorating this with copy is safer, and there might be some subtlety I am missing.

The behavior for `ArrayImageSource` is slightly different, because it is already cached (backed by an array already).

File: src/aspire/source/image.py
Comment (vineetbansal) [reply src/aspire/source/image.py:102]: docstring added

File: src/aspire/source/image.py
Comment (vineetbansal) [reply src/aspire/source/image.py:633]: The only scenario I can think of is if one wants to compare the performance of an uncached `src` vs. a cached version of it, without having to recreate it.

I don't know the typical use case of these methods, so happy to take it off. (of course feel free to tweak this branch directly).

**Comment by vineetbansal** [comment]
For the detection of `is_finalized` in `set_metadata` that we were talking about - currently the constructor of `ImageSource` is not calling the `update()` method. (in fact no code is using it, since it's new). So it seems reasonable to me to consider that as the safe method going forward, while `set_metadata` remains the mutating method and has to be used with caution. Constructors/other helper methods are thus allowed to use `set_metadata` if object construction is in progress.

**Comment by garrettwrong** [comment]
Besides updating the tutorials, it occurred to me that we should update the getting started documentation to highlight the syntax.

I will take a stab at resolving these merge conflicts to push this along.

**Comment by garrettwrong** [comment]
Working on finishing the PR now.

Documentation/gallery is rebuilding.  Thinking on `is_finalized`a bit more, we can probably just use `_mutable` for a clearer name.  I did go through and actually begin using `update` so the documentation is the API implemented in the PR and am working on the `_mutable` flag now.

I'd like to merge everything that is going into `develop` before running the pipeline experiments since that could take days to complete.

File: src/aspire/source/image.py
Comment (garrettwrong) [reply src/aspire/source/image.py:633]: I removed it and manually reverted the associated invocations.

File: gallery/tutorials/aspire_introduction.py
Comment (janden) [inline gallery/tutorials/aspire_introduction.py]: This probably doesn't do what you expect…

File: gallery/tutorials/tutorials/preprocess_imgs_sim.py
Comment (janden) [inline gallery/tutorials/tutorials/preprocess_imgs_sim.py:60]: Heh. Thanks!

File: gallery/tutorials/tutorials/preprocess_imgs_sim.py
Comment (janden) [inline gallery/tutorials/tutorials/preprocess_imgs_sim.py]: This can be moved down closer to where it's relevant (L150 or so).

File: gallery/tutorials/tutorials/preprocess_imgs_sim.py
Comment (janden) [inline gallery/tutorials/tutorials/preprocess_imgs_sim.py:238]: Nice!

File: src/aspire/storage/starfile.py
Comment (janden) [inline src/aspire/storage/starfile.py]: Just curious, but why is this needed?

File: tests/test_coordinate_source.py
Comment (janden) [inline tests/test_coordinate_source.py]: Does this still mutate?

File: tests/test_simulation_metadata.py
Comment (janden) [inline tests/test_simulation_metadata.py:78]: Why? Seems like a dangerous way to specify missing values…

File: tests/test_simulation_metadata.py
Comment (janden) [inline tests/test_simulation_metadata.py:102]: I'm confused. Why are we doing this?

File: src/aspire/source/image.py
Comment (janden) [inline src/aspire/source/image.py]: `ValueError` or just `RuntimeError`? Also below.

File: src/aspire/source/image.py
Comment (janden) [inline src/aspire/source/image.py:493]: Describe briefly what's going on here? Some sort of iterable check, otherwise we broadcast?

File: src/aspire/source/image.py
Comment (janden) [inline src/aspire/source/image.py:479]: Is `values` supposed to be one- or two-dimensional (the fact that `metadata_fields` can be a vector suggests both, but docstring is not clear here).

File: src/aspire/source/image.py
Comment (janden) [inline src/aspire/source/image.py:528]: Would it make sense to work with `MaskedArray`s here? Maybe something to explore in an issue.

File: src/aspire/source/image.py
Comment (janden) [inline src/aspire/source/image.py:896]: This only needs to access `n` from `self`, so we may want to consider making this a static/class/external method (it doesn't really belong to the class).

File: src/aspire/source/image.py
Comment (janden) [inline src/aspire/source/image.py]: 42?

**Review by janden — CHANGES_REQUESTED** [review]
Looks great! I just have some questions and suggestions but this should basically be good to go.

File: gallery/tutorials/tutorials/preprocess_imgs_sim.py
Comment (garrettwrong) [reply gallery/tutorials/tutorials/preprocess_imgs_sim.py:238]: Yeah after a tweak to whiten that actually worked out nicely. (Up to the linter restyling it not to my taste).

File: gallery/tutorials/tutorials/preprocess_imgs_sim.py
Comment (garrettwrong) [reply gallery/tutorials/tutorials/preprocess_imgs_sim.py]: Moved

File: gallery/tutorials/aspire_introduction.py
Comment (garrettwrong) [reply gallery/tutorials/aspire_introduction.py]: It does not, my oversight. Thanks for catching. Changed.

File: src/aspire/storage/starfile.py
Comment (garrettwrong) [reply src/aspire/storage/starfile.py]: I think there may be related changes in the other PR (911).

File: tests/test_coordinate_source.py
Comment (garrettwrong) [reply tests/test_coordinate_source.py]: Unless I have missed something, cache should i) generate images ii) reset only the generation_pipeline belonging to this source.  No metadata or image data should be mutated afaik.

It is true the pipeline reset is a mutation, but we will be recalling the array backed images, and so there should be no net effect to anything outside of `.images`.

It is probably true that in the immutable framework we no longer need the `self.generation_pipeline_forward(...)` line when loading ImageSource._cached_im images.  Anything adding to the pipeline after a cache call should yield a new source.  I can rm that or create an issue (if we continue this path)

(Another path.) Vineet had converted most of the cache calls to the immutable style for us and I reverted those because I didn't feel like caching was a transform.  If you are worried about it, or feel differently, I can quickly revert reverting the cache calls, so, making them to the immutable syntax. I think its just one commit and a couple missed lines from merges.

Let me know which way you prefer and I'll get it done. Thanks

File: tests/test_simulation_metadata.py
Comment (garrettwrong) [reply tests/test_simulation_metadata.py:78]: I agree with your other remark that masked arrays would be more appropriate.  Assuming we are not actually using these in practice, it is probably safe to make a followup issue to cleanup.  I'm not sure users would really notice, and I don't know if I can rush it out cleanly.

Comment (janden) [reply tests/test_simulation_metadata.py:78]: Agreed. Let's make an issue.

File: tests/test_simulation_metadata.py
Comment (garrettwrong) [reply tests/test_simulation_metadata.py:102]: I think because the table is a table of objects (arrays).

This test should just be checking that when we ask for the entire table we get the same thing as concatenating asking for each column (metadata key) .  A list could have been used instead of a vstack.  Not sure if that answers your concern.

File: src/aspire/source/image.py
Comment (garrettwrong) [reply src/aspire/source/image.py]: i replaced all the asserts with errors

File: src/aspire/source/image.py
Comment (garrettwrong) [reply src/aspire/source/image.py:479]: I think values can be 0, 1, or 2d.

Looking closer now, Im not sure the behavior when using multiple fields in conjunction with values would work for the dense case (where we dont want to assign the same values to each field.).  Will look a little more.

File: src/aspire/source/image.py
Comment (garrettwrong) [reply src/aspire/source/image.py:528]: Yes, I think so.

Comment (garrettwrong) [reply src/aspire/source/image.py:528]: https://github.com/ComputationalCryoEM/ASPIRE-Python/issues/920

File: src/aspire/source/image.py
Comment (garrettwrong) [reply src/aspire/source/image.py:896]: Converted to static method.   This and related methods could potentially be collapsed now that we have the metadata a dict already. I'll make an issue.

Comment (garrettwrong) [reply src/aspire/source/image.py:896]: https://github.com/ComputationalCryoEM/ASPIRE-Python/issues/921

File: src/aspire/source/image.py
Comment (garrettwrong) [reply src/aspire/source/image.py]: I htink just double oversight, I'll try removing and using the default

File: src/aspire/source/image.py
Comment (janden) [reply src/aspire/source/image.py]: Thanks!

File: src/aspire/source/image.py
Comment (janden) [reply src/aspire/source/image.py:896]: Thanks.

File: src/aspire/source/image.py
Comment (janden) [reply src/aspire/source/image.py:479]: Ok. Maybe just make an issue and leave this for now (from what I understand, the behavior is unchanged from before).

File: tests/test_coordinate_source.py
Comment (janden) [reply tests/test_coordinate_source.py]: Agreed that caching is not a transform, but it might still be confusing that all other operations on sources yield a new source but this one does not. I think it's a simpler rule of thumb to have all operations (whether they mutate or not, strictly speaking) generate a new source. This is a cached *view* of the source, if you will.

File: src/aspire/storage/starfile.py
Comment (janden) [reply src/aspire/storage/starfile.py]: Ok got it. Looks like it goes away there.

File: tests/test_simulation_metadata.py
Comment (janden) [reply tests/test_simulation_metadata.py:102]: Ok got it.

File: src/aspire/source/image.py
Comment (garrettwrong) [reply src/aspire/source/image.py:479]: Sure, https://github.com/ComputationalCryoEM/ASPIRE-Python/issues/924, thanks.

File: tests/test_coordinate_source.py
Comment (garrettwrong) [reply tests/test_coordinate_source.py]: No worries, I'll put that back, just a few minutes.

Comment (garrettwrong) [reply tests/test_coordinate_source.py]: Reverted. My bad on that one.

File: tests/test_simulation_metadata.py
Comment (garrettwrong) [reply tests/test_simulation_metadata.py:78]: https://github.com/ComputationalCryoEM/ASPIRE-Python/issues/920

File: tests/test_coordinate_source.py
Comment (janden) [reply tests/test_coordinate_source.py]: Totally fine. I see what you were trying to do. Just think it's simpler this way.

File: src/aspire/source/image.py
Comment (garrettwrong) [reply src/aspire/source/image.py:493]: Added too comments

```
+        # Check if we're an iterable, and in case we're not, broadcast
+        # the single values into a list `indices` long.
         try:
             iter(values)
         except TypeError:
             values = [values] * len(indices)
         else:
+            # Special case for single `str`, which are iterable, but
+            # need to be broadcast like a singleton.
             if isinstance(values, str):
                 values = [values] * len(indices)
```

File: src/aspire/source/image.py
Comment (garrettwrong) [reply src/aspire/source/image.py]: Using default of `None`, we did not raise, so I'd like to think that means we're not using this feature.

File: src/aspire/source/image.py
Comment (janden) [reply src/aspire/source/image.py:493]: Ok great. Thanks.

File: src/aspire/source/image.py
Comment (janden) [reply src/aspire/source/image.py]: Cool.

**Review by janden — APPROVED** [review]
Looks like we're good here. Nice work!

**Review by garrettwrong — APPROVED** [review]
Great. I'll merge it once the CI clears. Then start retesting the pipelines now that all v0.11.0 merges have occurred.  Hopefully just revisions to the release notes now. Thanks!

**Pull request merged**
